File: find-k-closest-elements.js
// Time Complexity : O(logn + k)
// Space Complexity : O(1) or O(k) depending on if you consider the output array as space
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * @param {number[]} arr
 * @param {number} k
 * @param {number} x
 * @return {number[]}
 */
var findClosestElements = function(arr, k, x) {
    // Base cases
    if (!arr || !arr.length) return [];
    if (x < arr[0]) return arr.slice(0, k);
    if (x > arr[arr.length - 1]) return arr.slice(arr.length - k, arr.length);
    
    let left = 0,
        right = arr.length - 1,
        mid;

    // Do a binary search for x over the WHOLE array
    while (left <= right) {
        mid = Math.floor((left + right) / 2);
        if (arr[mid] == x) break;
        else if (arr[mid] > x) right = mid - 1;
        else left = mid + 1;
    }
    
    // NOW set my window once I'm as close as possible to X
    left = Math.max(0, mid - k - 1);
    right = Math.min(arr.length - 1, mid + k - 1);

    // This while loop walks my left and right pointers away
    // from the found mid step by step, ensuring that they
    // don't go over the bounds of the array
    while (right - left > k - 1) {
        if (left < 0 || x - arr[left] <= arr[right] - x) right--;
        else if (right > arr.length - 1 || x - arr[left] > arr[right] - x) left++;
    }
    
    // Now just return a subarray from left to right
    return arr.slice(left, right + 1);
};

File: FindKClosestElements.cs
// Time Complexity : O(log(n-k))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
public IList<int> FindClosestElements(int[] arr, int k, int x) {
    IList<int> result = new List<int>();
    
    if(arr == null || arr.Length == 0)
        return result;
    
    //we low from satrt and higher from n-k
    int low = 0;
    int n = arr.Length;
    int high = n-k;
    
    while(low < high){
        int mid = (low+high)/2;
        
        //we are calculating distance from start and end
        //so that instead of moving one pointer, we can decide what will be our whole window
        int distStart = x - arr[mid];
        int distEnd = arr[mid+k] - x;
        
        //we compare distance from start with distance from end
        //so that high and low move closer
        if(distStart > distEnd)
            low = mid+1;
        else
            high = mid;
    }
    
    for(int i = low; i < low + k; i++)
    {
        result.Add(arr[i]);
    }
    
    return result;
}



//O(n)
public IList<int> FindClosestElements(int[] arr, int k, int x) {
    IList<int> result = new List<int>();
    if(arr == null || arr.Length == 0)
        return result;

    //Two pointers, one at start and one at end
    int left = 0;
    int right = arr.Length - 1;
    //indexes of start and end to see if it equals to k
    while(right - left + 1 > k)
    {
        //Subtract x from both start and end element
        int distLeft = x - arr[left];
        int distright = arr[right] - x;
        //Whichever result is larger, move that pointer
        //If difference of start and end from x is same, then we moved end element, 
        //because requirement is we need to take smaller element as the array is sorted.
        if(distLeft > distright)
            left++;
        else
            right--;
    }

    for(int i = left; i <= right; i++)
    {
        result.Add(arr[i]);
    }
    return result;
}